hash.c

```diff
@@ -303,6 +303,19 @@ objid_hash(VALUE obj)
  *
  * Certain core classes such as Integer use built-in hash calculations and
  * do not call the #hash method when used as a hash key.
+ *
+ * If you're implementing a #hash value method for a compound set of values in
+ * your own classes, best practice is to combine the class and the values by
+ * creating an array of them and then using the #hash method of the array, which
+ * is written to combine the hash values of its elements efficiently and
+ * safely. Manually combining hash values using mathematical operators is not
+ * recommended, as this is slower and more error-prone.
+ *
+ * For example:
+ *
+ *   def hash
+ *     [self.class, a, b, c].hash
```

Comment:
As mentioned on this issue, I don't think we should include the class here, and doing so is measurably slower.
In the vast majority of cases (where `#hash` is defined on a leaf class) there is no need, and the rest of cases (`#hash` defined in class which has subclasses) one should think what makes sense based on what they use those subclasses for.

I'd be fine to mention something like (and removing other references to the class):
> Consider including `self.class` in the values if the class may have subclasses and subclasses instances with the same values should not be considered equal.

Reply:
I don't get why it's about subclasses?

As a use-case, imagine a compiler IR where you have `Add` and `Sub` nodes, both of which refer to the same inputs, but have different operations. If they both have `x` and `y` as the hash values they'll collide in for example a GVN hash.

Reply:
> I don't get why it's about subclasses?

Because I'm thinking both `eql?` and `hash`, either:
* care about the exact class (e.g., `self.class == other.class` in `eql?`)
* don't care about the exact class (e.g., `other.is_a?(SomeClass)` in `eql?`)
(otherwise one of the two methods is broken)

You are right in your example that including the class makes sense and that avoids extra hash collisions.
I forgot about the case of 2 unrelated classes, with different hash methods but hash values accidentally collide, which semantically doesn't matter if `eql?` checks it, but performance-wise it can matter, that's a good point.
Another example is e.g. Complex and Rational, both might compute the hash as `[value1, value2].hash` but could cause unfortunate collisions in a Hash which can have both as keys.

As a counter-example, imagine a slight variation, where there is a `Add` node and a `AddOne` node (`AddOne < Add`), and I want `Add(x, 1).eql?(AddOne(x)) => true`, then it's the second case above.

Based on that I agree it seems best in general to include the class in the hash and recommend it here. For the Add/AddOne case above it should be clear enough that `[Add, x, y].hash` + `is_a?(Add)` should be used for that special case (and that's actually better than `[x, y].hash` + `is_a?(Add)` if there can be non-Add keys in the same Hash).